### ISSUE TITLE
[New build - DEFAULT] 20w26d: Steam Workshop Integration

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#2551]
The workshop is now integrated to the game. You can share your own creations and download other builders designs!
Still very much in beta test.

Changes
- Load blueprint menu now displays thumbnails. Please re-export your blueprint if you want to see the picture!
- Codex View Entity: overall layout updated.
- Codex View Entity: Description field added
- Codex View Entity: New "Share On Steam" button to upload a build to the workshop.

Bugfixes
#2546 #2534 Multiplayer broken.
#2542 NPC autopilot causes game fail. 
#2547 Rotator corruption can cause game freeze.

**Comment by tsunamayo** [comment]
The build and the workshop are now both live! Have fun posting on the workshop, I expect a few issue but we will sort them out!
Cheers

**Comment by TripleHelixxx** [comment]
Oooo workshop nice...  Bet this will make it easier for you to grab ships and stuff for promo. Easier for people to submit them as well.

**Comment by ProPeach** [comment]
This is brilliant! Glad you were able to work it out with Steam.

Do you have plans for an increase in the character count of an uploaded blueprints description, or the ability to upload a custom thumbnail image? Looking at the Workshops of other games likr SE, players really enjoy showing off their builds through those two means

**Comment by Kaiser-Indrasil** [comment]
Yeah, custom image thumbnails on the Workshop and increased character limit in the item's description both would help the presentation tremendously. Just look at what Aiwi's cooking up at Skywa.re and imagine that on the Steam Workshop:
![iketani](https://user-images.githubusercontent.com/63756057/85923596-4b7e8c80-b884-11ea-856b-099d1d88c5ee.gif)
Indeed, back when I played SE, I made sure that my designs looked presentable, too:
https://steamuserimages-a.akamaihd.net/ugc/1008185060158655993/5265C129219961C45A505B4D51B86EC996A8FFCD/

**Comment by Kaiser-Indrasil** [comment]
I hope we could have a better directory structure, too. Having each blueprint in its own separate folder would help keep the whole environment tidy and easy to find stuff in.
![image](https://user-images.githubusercontent.com/63756057/85931850-29563000-b8bf-11ea-9f60-4d84ae4b20c8.png)

**Comment by tsunamayo** [comment]
Yes I tried 16/9 but I had some issue with, I will give it another shot. I create a local folder for blueprints image but I upload straightaway. What is the upload process on space engineer?

**Comment by Kaiser-Indrasil** [comment]
There are 3 options:
* You can save to cloud and have access to your private blueprint collection even if you log in with your Steam account on another computer.
* If you've got cloud disabled, you can either export locally...
* ...or upload to the Workshop straight away.

The second option to export locally allows you to swap the default thumbnail (which is a simple screenshot from the game with UI removed). Then you launch the game again, and upload the updated blueprint to the Workshop from there.

Updating blueprints in Space Engineers is a pain though since you have to be in posession of the original local blueprint you uploaded from. In comparison, you already made it so much easier in Starship Evo by including the Reload/Update button.

**Comment by tsunamayo** [comment]
So what I could do is to pause the upload to two part, one it create the local folder, then you can replace your image if you want, then finally upload to steam with potentially your new custom image.

**Comment by Kaiser-Indrasil** [comment]
Exactly. Since I don't know of any other way to alter the thumbnail image, that would be the way to do it.

**Comment by 5URG3** [comment]
Hi tsunamayo,

I'm 5URG3, the creator of the Type 34 Merchantman workshop item you can see in the images above. I've uploaded over 100 submissions to the Space Engineers Steam Workshop over the last 5 years and have over 2000 followers. The majority of my creations have been popular enough to reach the top eight, the most popular uploads of the week, otherwise known as the “Front Page”

https://steamcommunity.com/profiles/76561198006667695/myworkshopfiles/

The workshop has proved to be a huge selling point for Space Engineers and the ability to edit thumbnails is critical to a vibrant workshop. **I strongly agree that players should be able to edit their thumbnails before uploading.**

In Space Engineers we can edit the blueprint thumbnail by manually locating the local blueprint folder and using a graphical editor to edit the PNG file “thumb”. This method is adequate for the task.

A potentially better way would be to have an in-game option to select their custom thumbnail using a File Explorer style UI.

As for the in-game description upload, I would say that this is far less important as the Steam client allows the editing of the workshop item description. The Steam Workshop uses BBcode for bold, italic and other styles and this is much easier to edit in a dedicated word processor. Pictures are also often added with links, this is again much easier in a word processor.

I am very impressed with Starship EVO so far and I am looking forward to future updates.

**Comment by tsunamayo** [comment]
@5URG3, thanks a lot for the detailed feedback.
I really like your builds, cant wait to see what you are doing in Starship EVO!
I will implement solution 1 (a manual replace) for the time being.
For screenshots, you mean in-game screenshot or some dedicated Preview screenshot?
I can also made them public by default - is it standard practice in other game? I only had a look at Scrap Mechanic and it was private, but thats an easy change.

**Comment by Kaiser-Indrasil** [comment]
I would personally advise against setting the default share setting to *Public* and leave it on **Hidden** instead. The reasons being, as a creator, you can add additional images to the item's Workshop page only after you've uploaded it. On top of that, oftentimes an item's description that looks perfectly fine and tidy in the game's text editor, but will be badly formatted on the item's Workshop page, leading to additional work.

The **Hidden** status prevents that since you can peacefully add your images and make sure the description looks alright before making it Public.

I agree that many people forget to set the item's share settings to Public but that can be remedied by simply reminding them to do it in a form of a pop-up message, very much like the one already in the game. In fact, I believe that on top of the original "Upload complete", the best solution would be to add a hint, something like: 

"Remember to set the Visibility to Public on the item's Workshop page."

**Comment by 5URG3** [comment]
Thanks, I'm currently doing my best to create something worthy of the games fantastic building system.

Solution 1 will do nicely, thank you.

Regarding screenshots, the default Steam screenshot (F12) is taken using the in-game resolution, typically 1920 x 1080. What I am suggesting is to include an option to take a high resolution screenshot at either double or quadruple the in-game resolution (3840 x 2160 or 7680 x 4320). This could be an option in photo mode. At this early stage of the games development I understand this is not a priority, but it would be a useful feature in the long term.

Regarding the visibility setting, my experience is mostly with Space Engineers. All SE workshop uploads are public by default. For most creators, the purpose of the Workshop is to show off their creations, therefore this visibility setting makes the most sense. More people will switch from Hidden to Public than would switch from Public to Hidden.

As a counterpoint to Kaiser-Indrasil's suggestion above, the amount of time it takes to add existing screenshots and a pre-written description (from a word processor) to a Workshop upload is less than ten minutes. During that time, the item will have had very few views, likely less than 10. It is also possible for creators to set visibility to Hidden during this time if they want to. Personally I have never had a problem with visibility settings set to Public.

If people wish to use the in-game text editor to make Workshop descriptions, it may need to use BBCode as that is how formatting is managed on the Workshop. This seems an unnecessary use of resources given how much easier it is to make Workshop descriptions using a word processor.

**Comment by tsunamayo** [comment]
Okay the new workshop workflow is up. I also changed to format to 16/9, but havent updated the default thumbnail yet.
I also need to draw decals on the preview.
Default visibility is at public, we will see if people are not ok with that.

**Comment by Kaiser-Indrasil** [comment]
Perhaps a simple screenshot would look more appealing than a preview shot with grey background? The beautiful star system we've got would get a proper exposition. People love a good background.

**Comment by tsunamayo** [comment]
@Kaiser-Indrasil it is complicated, but I could always cheat - but I also dont want it to be distracting, the star is the blueprint...

**Comment by Vicomt** [comment]
OK, workshop default - you're never going to get it right applying a hardcoded default. can we please have to the option to upload and set visibility at the same time? I'm sure Empyrion did it that way, as we ALWAYS upload private, then set public when the BP is ready.
within weeks you'll find that people are competing for top-spots on the workshop (whether they are actually competing or not), and for those that like making splashes on the workshop (ourselves included), customisation is the key, having the power to upload private OR public is such a time-saver, and means we don't pollute the workshop like/dislike system with potentially unfinished blueprints.

sorry, rant over, but just changing private to public like that is..... a bit far IMO

**Comment by Vicomt** [comment]
Next - backgrounds. again, whatever you do, you're going to annoy people with backgrounds. starfields get too busy, blank colours can be nice, but usually need some sort of framing.

Best way IMO again, is to give us options - select either blank(coloured?)/starfield background, apply a frame over image maybe, then composite the preview into it before uploading. As with other good software design, allowing a preview and the option to go back and change before committing is good.

**Comment by Vicomt** [comment]
Lastly (sorry, but a reply to 5URG3 above) - ten minutes is all it takes. we've had BP's pushed so far down workshop because some idiot decided to thumbs down an unfinished, accidentally visible BP. Early dislikes KILL workshop items, and believe me, we've been in some workshop battles.

**Comment by 5URG3** [comment]
I was unaware that changing the workshop visibility in-game was an option. This is definitely the way forward if it is easy to implement.

@Vicomt - I am all too aware of the dangers of an early dislike, the SE workshop is no exception, however I would think those that downvote that early are targeting the author specifically. If so, this is an issue that can not be solved by default visibility settings. Nevertheless, the in-game visibility option you've suggested above definitely seems the best option.

I do remain confused as to why anyone would want to upload an unfinished BP to the workshop and then update later. I'm reasonably sure the seven day timer starts the moment the build is uploaded. Why would anyone want their build to spend that time hidden? This is based on anecdotal evidence though, if anyone has any real data I'd love to see it.

**Comment by Vicomt** [comment]
one main reason for uploading a BP to the workshop unfinished is when multiple people are working on it ;)
I and my bro are co-builders, sometimes he's the host, sometimes I am. In the past in Empyrion, we used the steam workshop to share the BP's between us, using the co-contributor tag, which meant that BP's we were working on, were upload as well as finished ships.

**Comment by tsunamayo** [comment]
Okay I will put a visibility setting to the windows. Cheers

**Comment by Kaiser-Indrasil** [comment]
Just a pretty neatly introduced Blueprint Publicity setting I saw today playing KSP
![KSP_x64_DjjRUBNFFM](https://user-images.githubusercontent.com/63756057/88498840-2896e380-cfbc-11ea-9e9a-9e6688b9b6bd.png)
